src/vision/status_quo/alan_builds_a_task_scheduler.md

```diff
@@ -0,0 +1,704 @@
+					}
+				default:
+				}
+			case <-ap.stopCh:
+				ap.doneCh <- struct{}{}
+				return
+			}
+		}
+	}()
+}
+
+func (ap *applier) stop() error {
+	select {
+	case ap.stopCh <- struct{}{}:
+	case <-time.After(5 * time.Second):
+		return errors.New("took too long to signal stop")
+	}
+	select {
+	case <-ap.doneCh:
+	case <-time.After(5 * time.Second):
+		return errors.New("took too long to receive done")
+	}
+	return nil
+}
+
+func (ap *applier) apply(req Request) (string, error) {
+	reqID := ap.requestIDGenerator.next()
+	respRx, err := ap.notifier.register(reqID)
+	if err != nil {
+		return "", err
+	}
+
+	select {
+	case ap.requestCh <- requestTuple{requestID: reqID, request: req}:
+	case <-time.After(ap.requestTimeout):
+		if err = ap.notifier.trigger(reqID, fmt.Sprintf("failed to schedule %d in time", reqID)); err != nil {
+			return "", err
+		}
+	}
+
+	msg := ""
+	select {
+	case msg = <-respRx:
+	case <-time.After(ap.requestTimeout):
+		return "", errors.New("apply timeout")
+	}
+
+	return msg, nil
+}
+```
+
+Not fully grokking Rust ownership model in async, Alan implements the following code, but faced with a bunch of compiler error messages:
+
+```rust
+use async_std::task;
+
+pub struct Applier {
+    notifier: notify::Notifier,
+    ...
+}
+
+impl Applier {
+    pub fn new(req_timeout: Duration) -> Self {
+        ...
+        Self {
+            ...
+            notifier: notify::Notifier::new(),
+            ...
+        }
+    }
+    ...
+
+    pub async fn start(&self) -> io::Result<()> {
+        task::spawn(apply_async(
+            self.notifier,
+            ...
+        ));
+        ...
+        Ok(())
+    }
+    ...
+
+
+pub async fn apply_async(
+    notifier: notify::Notifier,
+    ...
+) -> io::Result<()> {
+  ...
+```
+
+```
+error[E0507]: cannot move out of `self.notifier` which is behind a shared reference
+  --> src/apply.rs:72:13
+   |
+72 |             self.notifier,
+   |             ^^^^^^^^^^^^^ move occurs because `self.notifier` has type `Notifier`, which does not implement the `Copy` trait
+```
+
+After discussing with [Barbara][], Alan adds `Arc` to provide a shared ownership between async tasks:
+
+```rust
+use async_std::{sync::Arc, task};
+
+pub struct Applier {
+    notifier: Arc<notify::Notifier>,
+    ...
+}
+
+impl Applier {
+    pub fn new(req_timeout: Duration) -> Self {
+        ...
+        Self {
+            ...
+            notifier: Arc::new(notify::Notifier::new()),
+            ...
+        }
+    }
+    ...
+
+    pub async fn start(&self) -> io::Result<()> {
+        task::spawn(apply_async(
+            self.notifier.clone(),
+            ...
+        ));
+        ...
+        Ok(())
+    }
+    ...
+
+
+pub async fn apply_async(
+    notifier: Arc<notify::Notifier>,
+    ...
+) -> io::Result<()> {
+  ...
+```
+
+Alan is satisfied with the compilation success for the moment, but doesn't feel confident about the production readiness of Rust async.
```

Comment:
I don't think this part is related to async rust. It's just how rust, ownership system works. It is the same in sync rust. And it is not related to production readiness, if you use rust concepts and port it directly to go, it will be roughly the same.